mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp

```diff
@@ -309,6 +314,99 @@ emulatedVectorLoad(OpBuilder &rewriter, Location loc, Value base,
       newLoad);
 }
 
+/// Emits `memref.generic_atomic_rmw` op to store a subbyte-sized value to a
+/// byte in memory, with a mask. The `valueToStore` is a vector of subbyte-sized
+/// elements, with size of 8 bits, and the mask is used to select which elements
+/// to store.
+///
+/// Inputs:
+///   linearizedMemref = |a|b|c|d| : <4xi2> (<1xi8>)
+///   linearizedIndex = 2
+///   valueToStore = |e|f|g|h| : vector<4xi2>
+///   mask = |0|0|1|1| : vector<4xi1>
+///
+/// Result:
+///   linearizedMemref = |a|b|g|h| : <4xi2> (<1xi8>)
+static void atomicStore(OpBuilder &builder, Location loc,
+                        MemRefValue linearizedMemref, Value linearizedIndex,
+                        VectorValue valueToStore, Value mask,
+                        int64_t numSrcElemsPerDest) {
+  assert(valueToStore.getType().getRank() == 1 && "expected 1-D vector");
+  auto atomicOp = builder.create<memref::GenericAtomicRMWOp>(
+      loc, linearizedMemref, ValueRange{linearizedIndex});
+  Value origValue = atomicOp.getCurrentValue();
+
+  OpBuilder::InsertionGuard guard(builder);
+  builder.setInsertionPointToStart(atomicOp.getBody());
+
+  auto oneVectorType = VectorType::get({1}, origValue.getType());
+  auto fromElem = builder.create<vector::FromElementsOp>(loc, oneVectorType,
+                                                         ValueRange{origValue});
+  auto vectorBitCast =
+      builder.create<vector::BitCastOp>(loc, valueToStore.getType(), fromElem);
+
+  auto select =
+      builder.create<arith::SelectOp>(loc, mask, valueToStore, vectorBitCast);
+  auto bitcast2 = builder.create<vector::BitCastOp>(loc, oneVectorType, select);
+  auto extract = builder.create<vector::ExtractOp>(loc, bitcast2, 0);
+  builder.create<memref::AtomicYieldOp>(loc, extract.getResult());
```

Comment:
I think it is better to insert two comments because I need to parse the logic every time. Although they look dummy, they are human-friendly IMO. Also, I'd suggest renaming the variables like below, please take a look.

```suggestion

  // Construct the vector value from the scalar.
  auto oneElemVecType = VectorType::get({1}, origValue.getType());
  Value origVecValue = builder.create<vector::FromElementsOp>(loc, oneElemVecType,
                                                         ValueRange{origValue});
  origVecValue =
      builder.create<vector::BitCastOp>(loc, valueToStore.getType(), origVecValue);

  // Construct the final masked value and yield it.
  Value maskedValue =
      builder.create<arith::SelectOp>(loc, mask, valueToStore, vectorBitCast);
  maskedValue = builder.create<vector::BitCastOp>(loc, oneVectorType, select);
  Value scalarMaskedValue = builder.create<vector::ExtractOp>(loc, maskedValue, 0);
  builder.create<memref::AtomicYieldOp>(loc, scalarMaskedValue);
```